platform/fabric/core/generic/vault/vault.go

```diff
@@ -92,3 +111,108 @@ func (p *populator) Populate(rws *vault.ReadWriteSet, rwsetBytes []byte, namespa
 
 	return nil
 }
+
+type marshaller struct{}
+
+func (m *marshaller) Marshal(rws *vault.ReadWriteSet) ([]byte, error) {
+	rwsb := rwsetutil.NewRWSetBuilder()
+
+	for ns, keyMap := range rws.Reads {
+		for key, v := range keyMap {
+			if v.Block != 0 || v.TxNum != 0 {
+				rwsb.AddToReadSet(ns, key, rwsetutil.NewVersion(&kvrwset.Version{BlockNum: v.Block, TxNum: v.TxNum}))
+			} else {
+				rwsb.AddToReadSet(ns, key, nil)
+			}
+		}
+	}
+	for ns, keyMap := range rws.Writes {
+		for key, v := range keyMap {
+			rwsb.AddToWriteSet(ns, key, v)
+		}
+	}
+	for ns, keyMap := range rws.MetaWrites {
+		for key, v := range keyMap {
+			rwsb.AddToMetadataWriteSet(ns, key, v)
+		}
+	}
+
+	simRes, err := rwsb.GetTxSimulationResults()
+	if err != nil {
+		return nil, err
+	}
+
+	return simRes.GetPubSimulationBytes()
+}
+
+func (m *marshaller) Append(destination *vault.ReadWriteSet, raw []byte, nss ...string) error {
+	txRWSet := &rwset.TxReadWriteSet{}
+	err := proto.Unmarshal(raw, txRWSet)
+	if err != nil {
+		return errors.Wrap(err, "provided invalid read-write set bytes, unmarshal failed")
+	}
+
+	source, err := rwsetutil.TxRwSetFromProtoMsg(txRWSet)
+	if err != nil {
+		return errors.Wrap(err, "provided invalid read-write set bytes, TxRwSetFromProtoMsg failed")
+	}
+
+	for _, nsrws := range source.NsRwSets {
```

Comment:
We can do this more efficiently and concisely using sets
```
namespaces := collections.NewSet(nss...)

for _, nsrws := range source.NsRwSets {
 if !namespaces.Contains(nsrws.NameSpace) {
  continue
 }

 for _, read := range nsrws.KvRwSet.Reads {
 ...
```

Reply:
okay